### ISSUE TITLE
[Feature request]: next dev --turbo support

### ISSUE THREAD

**Issue by leerob** [serwist/serwist#54]
### Which project is this feature request for?

@serwist/next

### Describe the feature you'd like to request

It would be awesome to have a version of Serwist that works with turbopack.

### Describe the solution you'd like

https://turbo.build/pack/docs/features/customizing-turbopack#webpack-loaders-for-nextjs

### Describe alternatives you've considered

Webpack 😄

**Comment by DuCanhGH** [comment]
> It would be awesome to have a version of Serwist that works with turbopack.

I agree! However, it seems to be not really feasible at the moment. Since Turbopack isn't configurable through plugins yet, our only options to provide rudimentary support for Turbopack would be to either introduce a post-build script that the user runs manually, which would only work for `next build --turbo` (which doesn't seem to be a thing yet?) and would be really limited, for we are not able to access things the likes of webpack config, fully resolved Next.js config, etc., or write a Turbopack loader as suggested, but I'm not really sure as to when this loader is run. If it runs after the build script is done with static files such as ".next/static/chunks/**/*.js" then we can make it work. Otherwise, we may as well just wait for Turbopack's plugin support. I hope you understand the situation...

**Comment by bestickley** [comment]
@leerob, any updates on when Turbopack will support plugins?

**Comment by leerob** [comment]
Hey! It supports some today: https://nextjs.org/docs/app/api-reference/next-config-js/turbo#webpack-loaders

**Comment by bestickley** [comment]
@leerob, thanks for the response. My understanding of loaders vs plugins is that loaders only allow you to transform a single file whereas plugins allow customization over build process. See more [here](https://stackoverflow.com/questions/37452402/webpack-loaders-vs-plugins-whats-the-difference)). I think the next serwist plugin requires more customization over the build process than just modifying single files.

**Comment by mi-hol** [comment]
> I agree! However, it seems to be not really feasible at the moment.
> Since Turbopack isn't configurable through plugins yet,

Turbopack stated their development roadmap recently.

For this issue I'd propose to use the label "backlog"

**Comment by FleetAdmiralJakob** [comment]
Could serwist add a notice like this one from Sentry to the terminal if someone tries serwist with turbopack? This would make it obvious that they don't work together, and you wouldn't have to try to debug what's wrong if it is not working.

<img width="1205" alt="Image" src="https://github.com/user-attachments/assets/4961fc3a-fd25-4d12-8bb6-7fa16b832419" />

**Comment by DuCanhGH** [comment]
@FleetAdmiralJakob will do :)

**Comment by elie222** [comment]
Still get this warning when doing:

```ts
const withSerwist = withSerwistInit({
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  disable: env.NODE_ENV !== "production",
});
```

Warning:

```
[@serwist/next] WARNING: You are using '@serwist/next' with `next dev --turbopack`, but Serwist doesn't support Turbopack at the moment. It is recommended
      that you set `disable` to `process.env.NODE_ENV !== "production"`. Follow this issue for progress on Serwist + Turbopack: https://github.com/serwist/serwist/issues/54.
      You can also suppress this warning by setting SERWIST_SUPPRESS_TURBOPACK_WARNING=1.
```

**Comment by daveycodez** [comment]
> oduction"`. Follow this issue for progress on Serwist + Turbopack: https://github.com/serwist/serwist/issues/54.
>       You can also suppress this warning by setting SERWIST_SUPPRESS_TURBOPACK_WARNING=1.

Need to add this to env, but not ideal :/

SERWIST_SUPPRESS_TURBOPACK_WARNING="1"

**Comment by yasseralsaidi** [comment]
any update here?

**Comment by daveycodez** [comment]
```bash
Checked 72 files in 13ms. No fixes applied.
[@serwist/next] WARNING: You are using '@serwist/next' with `next dev --turbopack`, but Serwist doesn't support Turbopack at the moment. It is recommended
      that you set `disable` to `process.env.NODE_ENV !== "production"`. Follow this issue for progress on Serwist + Turbopack: https://github.com/serwist/serwist/issues/54.
      You can also suppress this warning by setting SERWIST_SUPPRESS_TURBOPACK_WARNING=1.
   ▲ Next.js 15.2.4 (Turbopack)
   - Local:        http://localhost:3000
   - Network:      http://192.168.50.137:3000
   - Environments: .env
   - Experiments (use with caution):
     · turbo

 ✓ Starting...
[@serwist/next] WARNING: You are using '@serwist/next' with `next dev --turbopack`, but Serwist doesn't support Turbopack at the moment. It is recommended
      that you set `disable` to `process.env.NODE_ENV !== "production"`. Follow this issue for progress on Serwist + Turbopack: https://github.com/serwist/serwist/issues/54.
      You can also suppress this warning by setting SERWIST_SUPPRESS_TURBOPACK_WARNING=1.
 ✓ Compiled in 226ms
[@serwist/next] WARNING: You are using '@serwist/next' with `next dev --turbopack`, but Serwist doesn't support Turbopack at the moment. It is recommended
      that you set `disable` to `process.env.NODE_ENV !== "production"`. Follow this issue for progress on Serwist + Turbopack: https://github.com/serwist/serwist/issues/54.
      You can also suppress this warning by setting SERWIST_SUPPRESS_TURBOPACK_WARNING=1.
 ✓ Ready in 766ms
[@serwist/next] WARNING: You are using '@serwist/next' with `next dev --turbopack`, but Serwist doesn't support Turbopack at the moment. It is recommended
      that you set `disable` to `process.env.NODE_ENV !== "production"`. Follow this issue for progress on Serwist + Turbopack: https://github.com/serwist/serwist/issues/54.
      You can also suppress this warning by setting SERWIST_SUPPRESS_TURBOPACK_WARNING=1.
```

I feel attacked

**Comment by daveycodez** [comment]
@DuCanhGH Plz undo the warning

**Comment by DuCanhGH** [comment]
@elie222 should be fixed in 9.0.13 :)

**Comment by soyricardodev** [comment]
Now that we have build support for nextjs with turbopack, can we recheck this issue status? i think should be possible to create a better workaround

**Comment by DuCanhGH** [comment]
@soyricardodev we can create a manual post-build script :) I suppose it won't have access to Next.js config though.

**Comment by dimitrov93** [comment]
SERWIST_SUPPRESS_TURBOPACK_WARNING="1"

worked for me

**Comment by jeremybosma** [comment]
Is this still being worked on/maintained?

**Comment by DuCanhGH** [comment]
@jeremybosma in https://github.com/serwist/serwist/discussions/278 I've detailed plans for supporting Turbopack, but I'm not able to work on that just yet. Sorry about that.

**Comment by DuCanhGH** [comment]
I've committed the current work on support for Turbopack @ [packages/turbo](https://github.com/serwist/serwist/tree/main/packages/turbo). It should tells you guys much about the plans for supporting it. I think it looks pretty good already! It's just that `next build --turbopack` doesn't allow me to load `esbuild` in `@serwist/turbopack`'s Route Handler at all, and that locating the route at `app/serwist/[path]/route.ts` would cause issues with the service worker's `scope`. Other than that, yeah, the current code should pretty much lay out how Turbopack support would become in Serwist 10.

**Comment by andr-ec** [comment]
@DuCanhGH looks like the 11 preview has a bug. the example project inside of the repo on preview 11 shows this error: 

```
Uncaught SyntaxError: Cannot use 'import.meta' outside a module (at sw.js:330:52)
```

**Comment by DuCanhGH** [comment]
@andr-ec I can't reproduce the issue at all. Weird...

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/6bd1dd7c-1247-42ec-973f-03a974ea9dde" />

We are bundling the service worker with the following options:

```
await (await esbuild).build({
  entryPoints: [{ in: config.swSrc, out: "sw" }],
  format: "esm",
  sourcemap: true,
  bundle: true,
  write: false,
  minify: !isDev,
  define: injectionPoint
    ? {
        [injectionPoint]: manifestString,
      }
    : undefined,
  outdir: process.cwd(),
  entryNames: "[name]",
});
```

`format` is set to `esm`, and the `serwist` package has `"type": "module"`, so, logically, `import.meta` should be available.

**Comment by PinkaminaDianePie** [comment]
Does it work with `/pages`? Can't make it work, get 404: `Failed to register a ServiceWorker for scope ('http://localhost:8000/') with script ('http://localhost:8000/serwist/sw.js'): A bad HTTP response code (404) was received when fetching the script.`

**Comment by DuCanhGH** [comment]
@elsigh the problem was that `esbuild-wasm` was in `dependencies`, causing a version mismatch between the user's and the package's `esbuild-wasm`, so Next.js ignored `serverExternalPackages` and tried to bundle it. Should be fixed in Preview 13!

**Comment by elsigh** [comment]
Preview 13 works without any special esbuild config in `createSerwistRoute`!

**Comment by oscar-b** [comment]
> [@oscar-b](https://github.com/oscar-b) you set `globDirectory` to `path.join(process.cwd(), '.next/static')`, so Serwist's default precache manifest transformer was not applied. Maybe try setting `globPatterns` to `[".next/static/**/*.{js,css,html,ico,apng,png,avif,jpg,jpeg,jfif,pjpeg,pjp,gif,svg,webp,json,webmanifest}"]` instead?

Ah yes I see, that works much better! 👍🏻 

> serwist Precaching 643 files. 2 files are already cached.


Another issue I'm seeing is that the request to `serwist/sw.js` works in dev, but generates a HTTP 404 in a prod build, really weird?


I was also looking at `defaultCache` and noticed some rules for `/api/`, however due to reasons (don't ask..) I have it prefixed like `/abc/api/`, should we just re-create the rules in this case or would it make sense to have that as a parameter?

**Comment by DuCanhGH** [comment]
@oscar-b

> but generates a HTTP 404 in a prod build, really weird?

Unless Next.js screws up, or the route fails to be pregenerated, that shouldn't happen. Try producing a reproduction.

> I have it prefixed like /abc/api/, should we just re-create the rules in this case 

Indeed!

> or would it make sense to have that as a parameter (like apiBase: '/abc/api/'

I'm thinking of making a function version of `defaultCache` that allows for customizing as well.

**Comment by DuCanhGH** [comment]
@oscar-b yeah that happens due to the code using `Iterator.map`, which is not supported by Node.js 20. I'll get to that and add tests soon!

**Comment by bolechen** [comment]
> [@yasseralsaidi](https://github.com/yasseralsaidi) currently the Preview would work well with Next.js, but I've been a bit too lazy and busy to finish up Serwist 10...

how to use the preview version, thx for your works.

**Comment by CrusaderGoT** [comment]
@Huh-David 

I have tried your approach.
The key was manually putting in the packages I wanted to install in package.json, and then running `npm install`.

But I am getting a new error.
```
[ERROR] The working directory "C:\Users\DELL\Documents\my-programming-folder\projects\achebestan" is not an absolute path

Error occurred prerendering page: "/_serwist/sw.js.map"
Error: Build failed.
  Error: The working directory "C:\Users\DELL\Documents\my-programming-folder\projects\achebestan" is not an absolute path
      at ignoreListedFrames
errors: [Array]
warnings: []

Export encountered an error on /_serwist/{path}/route: /_serwist/sw.js.map
Exiting the build.

Next.js build failed with exit code: 1 and signal: null
PS C:\Users\DELL\Documents\my-programming-folder\projects\achebestan>
```
Even though that path is an absolute path.

I have searched for solutions, and found none. Sorry for the continued requests

**Comment by Huh-David** [comment]
@CrusaderGoT Please try pnpm instead to make use of the pnpm lock file. Maybe npm got something wrong with the dependency versions.

**Comment by Huh-David** [comment]
@PerkyZZ999 If service workers are an important factor in your application and you value stability, you should definitely stick to the webpack version of Serwist with Next.js 15 for production until the turbopack package of Serwist becomes stable and well documented. 

@DuCanhGH Do you have any ideas when Serwist 10 will be released? :)

**Comment by DuCanhGH** [comment]
@Huh-David yeah I'm really sorry about Serwist 10 just stuck at Preview like that. I've frankly been quite busy and forgetting about Serwist...

@CrusaderGoT @rodrigonzalz please ensure that `serverExternalPackages` includes `"esbuild-wasm"`. Thanks! Also make sure that you guys are on Serwist 10 Preview 14 (`@serwist/turbopack@10.0.0-preview.14`).

**Comment by DuCanhGH** [comment]
Please refer to [the Serwist 10 example located at the v9.1 branch](https://github.com/serwist/serwist/tree/d772d9aaaa60b81b214d2c95f957aa713956372a/examples/next-turbo-basic) rather than the Serwist 9 example located at the main branch. Thank you!!!

**Comment by Taoister39** [comment]
> [@Taoister39](https://github.com/Taoister39) you have to use the latest Preview version, which is currently `@serwist/turbopack@10.0.0-preview.14`. A backport for Serwist 9 would be available soon.

I'm just referring to this repository...
Please check if next-turbo-basic example project can run.

**Comment by PerkyZZ999** [comment]
> [@DuCanhGH](https://github.com/DuCanhGH) - Looks like there is another vulnerability so 16.0.9 now
> 
> I'm currently waiting on this fix before using serwist. Otherwise will need to look at other libraries or do custom

Same here, I need to wait for a compatible Serwist version for 16.0.10 (another update was out again smh..)

@DuCanhGH Please update when you got times ! <3 💯

**Comment by DuCanhGH** [comment]
Serwist 9.3.1 adds the config mode, which both supports Turbopack and automatically precaches prerendered routes, to `@serwist/next`! This is an alternative to using `@serwist/turbopack` as is described above. See more at #314!

**Comment by FleetAdmiralJakob** [comment]
Hi @DuCanhGH, 

One question regarding `@serwist/turbopack`. In the comment for the `app/serwist/[path]/route.ts`, it says

> Copy the relevant Next.js configuration (assetPrefix,
    basePath, distDir) over if you've changed them.

What is all the relevant Next.js configuration? Is it only these three things or more? Can you maybe make this a little bit clearer?

**Comment by DuCanhGH** [comment]
@FleetAdmiralJakob I just spend some free time every day ;)

I personally find configurator mode way more intuitive too, but some may find not having to modify package.json... more convenient.

**Comment by Ripper1028** [comment]
Thanks Admiral, I'm new to js, and just started learning next js. I've been planning this project, and had originally planned on using php because I know it, but learned that php isn't a good choice for a pwa. All the AI bots just say use next-pwa, but when installing and seeing all the deprecated libraries I started looking for the replacement.

**Comment by svarup** [comment]
`process is not defined`, 

like to access `process.env.NEXT_PUBLIC_TEST`, in service worker give error

also for 

```
...
navigationPreload: true,
runtimeCaching: defaultCache,
importScripts: ['custom-sw.js'],
fallbacks: {
...
```

it gives

`Failed to execute 'importScripts' on 'WorkerGlobalScope': Module scripts don't support importScripts().`


both issues on tourbopack and worked on webpack

**Comment by nikelborm** [comment]
I have a folder with large audio samples in `public`. I used `next build --webpack` with the following config:
```ts
import withSerwistInit from '@serwist/next'

const withSerwist = withSerwistInit({
  // ...
  exclude: [
    // default exclusion:
    /\.map$/,
    /^manifest.*\.js$/,
    // Excludes everything in public/samples
    /^samples\/.*$/,
  ],
})
```

This option seems to be webpack-only:
https://github.com/search?q=repo%3Aserwist%2Fserwist+%2F%5Cbexclude%5Cb%2F+NOT+path%3A*tsconfig.json&type=code

I have migrated to the configurator mode. The `serwist build` currently produces a bunch of warnings like these: `public/samples/slow_strum/accord_G__strength_v.wav is 2.12 MB, and won't be precached. Configure maximumFileSizeToCacheInBytes to change this limit`. While developing, I didn't really care to re-encode and compress them into more web-suitable formats <ins>yet</ins>. I don't want to rely on the assumption that the samples will always be larger than a certain size and that all other artifacts I want precached would be below it.

The alternative now seems to be manually editing the resulting `sw.js`, but I haven't yet verified whether it's enough or whether other artifacts need to be edited or removed.

Is there a better way to achieve this with either Turbopack or configurator mode? Should I open a new issue about this?
